flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SingleLogicalSlot.java

```diff
@@ -169,23 +168,19 @@ public void release(Throwable cause) {
 		releaseFuture.complete(null);
 	}
 
-	private CompletableFuture<?> signalPayloadRelease(Throwable cause) {
+	private void signalPayloadRelease(Throwable cause) {
 		tryAssignPayload(TERMINATED_PAYLOAD);
-		payload.fail(cause);
-
-		return payload.getTerminalStateFuture();
+		payload.failAsync(cause); //TODO this goes from the pool against the execution, has to by sync'ed back into the jm main thread
 	}
 
 	private void returnSlotToOwner(CompletableFuture<?> terminalStateFuture) {
-		final CompletableFuture<Boolean> slotReturnFuture = terminalStateFuture.handle((Object ignored, Throwable throwable) -> {
+		terminalStateFuture.handle((Object ignored, Throwable throwable) -> {
 			if (state == State.RELEASING) {
 				return slotOwner.returnAllocatedSlot(this);
 			} else {
 				return CompletableFuture.completedFuture(true);
 			}
-		}).thenCompose(Function.identity());
-
-		slotReturnFuture.whenComplete(
+		}).whenComplete( //TODO this could be inside the job master main thread, should be inside slot pool main thread
```

Comment:
I think this change breaks the semantics of the `releaseFuture`. By not composing we will execute the `whenComplete` when `slotOwner.returnAllocatedSlot` returns a future and not when this future has been completed. Thus, I think we should revert this change and add a test which guards this behavior.

Reply:
Good catch, yes that was a clear mistake :-(

Reply:
Added a test.